### PR TITLE
Simplify extents in ops

### PR DESCRIPTION
**Pull request by jacobhinkle**
This PR uses `SimplifyingIrBuilder::maybeCastExpr` to simplify scalars in `maybeCastOp`, which is used in ops to simplify extents and other scalars at definition. This prevents introducing new cast expressions when those scalars are constant, instead just creating a new constant with the same value but a new dtype. This has a big effect when we use a constant `Int` as an `Index`, since those casts generally prevent some simplifications.

As an example, the incomplete fusion from #963 used to print like this
```
Inputs:                                                                                                                                  
  T0_g[ iS0{i0}, iS1{i2} ], float                                   
Outputs:                                                                                                                                                                                                                                                                          
%kernel_math {                                                                                                                           
T1_l[ iS2{( (nvfuser_index_t)(9) )}, iS3{( (nvfuser_index_t)(5) )} ]                                                                     
   = Set( T0_g[ iS0{i0}, iS1{i2} ], cache_op=Streaming )            
i7 = (nvfuser_index_t)(5);                                                                                                               
i16 = i7 + 2;                                                                                                                            
i22 = i7 + 3;                                                                                                                            
i35 = i7 - 2;                                                                                                                            
i24 = fmin(i7, 3);                                                                                                                       
i26 = fmax(2, i24);                                                                                                                      
i29 = -i7;                                                                                                                               
i31 = i26 + i29;
i37 = i35 + i31;                                                                                                                         
Resize: iS5{( (nvfuser_index_t)(5) )}rf by -2 and ( ( fmax(2, ( fmin(( (nvfuser_index_t)(5) ), 3) )) ) + ( -( (nvfuser_index_t)(5) ) ) ) 
-> bS6{( ( ( (nvfuser_index_t)(5) ) - 2 ) + ( ( fmax(2, ( fmin(( (nvfuser_index_t)(5) ), 3) )) ) + ( -( (nvfuser_index_t)(5) ) ) ) )}rf
i5 = (nvfuser_index_t)(9);                                                                                                               
T2_l[ iS4{( (nvfuser_index_t)(9) )}, bS6{( ( ( (nvfuser_index_t)(5) ) - 2 ) + ( ( fmax(2, ( fmin(( (nvfuser_index_t)(5) ), 3) )) ) + ( -(
 (nvfuser_index_t)(5) ) ) ) )}rf ]                                                                                                       
   = slice( T1_l[ iS2{( (nvfuser_index_t)(9) )}, iS3{( (nvfuser_index_t)(5) )} ], { {0, i5, 1} {2, i26, 1} } )                           
}
```
and after this PR it now prints like this
```
Inputs:
  T0_g[ iS0{i0}, iS1{i2} ], float
Outputs:

%kernel_math {
T1_l[ iS2{9}, iS3{5} ]
   = Set( T0_g[ iS0{i0}, iS1{i2} ], cache_op=Streaming )
Resize: iS5{5}rf by -2 and -2 -> bS6{1}rf
T2_l[ iS4{9}, bS6{1}rf ]
   = slice( T1_l[ iS2{9}, iS3{5} ], { {0, 9, 1} {2, 3, 1} } )
}
```

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
`ExpandedBroadcastGlobalIntermediateTest.TheTest_CUDA` is failing. It appears to be trying to bind the expanded extent instead of the true extent. I am not sure yet why this was succeeding before this PR, but the offending line appears to be https://github.com/NVIDIA/Fuser/blob/main/csrc/expr_evaluator.cpp#L128.

**Comment by jacobhinkle** [comment]
> `ExpandedBroadcastGlobalIntermediateTest.TheTest_CUDA` is failing. It appears to be trying to bind the expanded extent instead of the true extent. I am not sure yet why this was succeeding before this PR, but the offending line appears to be https://github.com/NVIDIA/Fuser/blob/main/csrc/expr_evaluator.cpp#L128.

Actually I think this line should change to copying the value: https://github.com/NVIDIA/Fuser/blob/4f6ee3d491c17d33c757ec6d172246eae0299e35/csrc/ir/builder.cpp#L276.

**Comment by zasdfgbnm** [comment]
Is the executor allocating a wrong output tensor?

**Comment by jacobhinkle** [comment]
> Is the executor allocating a wrong output tensor?

It's actually an intermediate global tensor. The allocator is correct but I think the binding code is incorrect when we have expanded extent. We should instead check that the input has an expanded axis or a broadcast in that dimension. If we check for expanded axis then we'll also need to update the allocator to expand the needed dims also. Will play with it in a couple hours.

**Comment by jacobhinkle** [comment]
According to this we don't need to worry about broadcast expansion for intermediate tensors https://github.com/NVIDIA/Fuser/blob/main/csrc/executor.cpp#L558-L563

But I think we do since we are now using the same mechanism to bind extents/maybeExpandedExtents as we use for inputs and outputs.

File: csrc/executor.cpp [hunk 1]
```diff
@@ -1203,8 +1203,7 @@ std::vector<FusionExecutor::GlobalBufferInfo> FusionExecutor::
     GlobalBufferInfo info;
     info.tv = tv;
     info.zero_init = alloc->zeroInit();
-    std::tie(info.sizes, info.strides) =
-        inferShapeOfIntermediate(tv, alloc, expr_eval);
+    std::tie(info.sizes, info.strides) = inferShapeOfOutput(tv, expr_eval);
```
Comment (jacobhinkle) [inline csrc/executor.cpp:1206]: `inferShapeOfIntermediate` is no longer used. I will remove it if we go this direction.

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
> The allocator is correct

The allocator allocated the proper amount of elements but I changed it to also expand the intermediate tensor if needed.

**Comment by naoyam** [comment]
I'm not 100% sure if this is completely safe. Looks like we are omitting `(nvfuser_index_t)` in, e.g., `(nvfuser_index_t)2**30`. Suppose `nvfuser_index_t` is `int64_t`. What happens if, say, there's a multiplication of such large values? For example, currently we should generate `((nvfuser_index_t)2**30) * 2`, which should not overflow, but doesn't it overflow if the cast is omitted, i.e., `(2**30) * 2`?

Maybe we should always explicitly add a suffix for integer values? https://github.com/NVIDIA/Fuser/blob/main/csrc/codegen.cpp#L177

**Comment by zasdfgbnm** [comment]
@naoyam Another issue of this PR might be it does not work with shift. I remembered that at some point I changed the allocation of intermediate tensors to use shape instead of `kir:: Allocate`, all tests pass, but compute sanitizer reported real errors, and you eventually reverted that part.

**Comment by jacobhinkle** [comment]
> Maybe we should always explicitly add a suffix for integer values? https://github.com/NVIDIA/Fuser/blob/main/csrc/codegen.cpp#L177

Yeah `2L ** 30L` would not overflow right? I can add that in a separate small pr.

**Comment by naoyam** [comment]
> @naoyam Another issue of this PR might be it does not work with shift. I remembered that at some point I changed the allocation of intermediate tensors to use shape instead of `kir:: Allocate`, all tests pass, but compute sanitizer reported real errors, and you eventually reverted that part.

@jacobhinkle Could you please split out the allocation of intermediate tensors as a separate PR? That would need some workaround for shift, which I'll take care of it. The other changes seem fine and seems helpful for other issues.

File: test/test_resize.cpp [hunk 2]
```diff
@@ -3115,4 +3115,50 @@ TEST_F(NVFuserTest, dynamicReshapeIssue1393) {
   testValidate(fusion, outputs, {t0, t1}, {ref}, __LINE__, __FILE__);
 }
 
+// Test that we can slice a trivially expanded tensor to size 1 then squeeze
+// See https://github.com/NVIDIA/Fuser/issues/963
```
Comment (jacobhinkle) [inline test/test_resize.cpp:3119]: This test should've been added in #975, but I'm including it here since it showcases the simplification.

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
There is only one test failure, `ExpandedBroadcastGlobalIntermediateTest.TheTest_CUDA`, and #1469 will fix it.

**Comment by jacobhinkle** [comment]
!build

**Pull request merged**